File: 백준/Gold/1106. 호텔/호텔.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

#setrecursionlimit(10**9)
input = stdin.readline

C, N = map(int, input().split()) # C - 적어도 얻고자 하는 이득 이 때 최소 비용 구하기
hotel=[[0,0]] #cost, benefit
for i in range(N):
    line= list(map(int, input().split()))
    hotel.append(line)

maxCost = float('INF')
dp= [maxCost for _ in range(C+100)]
dp[0]=0

for i in range(1, N+1):
    cost = hotel[i][0]
    benefit = hotel[i][1]
    
    for j in range(benefit, C+100):    
        dp[j] = min(dp[j], dp[j-benefit]+cost)
        
# print(dp)     
print(min(dp[C:]))

File: 백준/Gold/1106. 호텔/호텔.swift
let cn = readLine()!.split{$0 == " "}.map{Int($0)!}

let (c,n) = (cn[0],cn[1])

var arr: [(cost:Int, customer:Int)] = []

var dp: [Int] = [Int](repeating: 200000, count: c+101) // dp[i] = i 명을 모을 때 최소 비용

for _ in 0..<n {
    let info = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    arr.append((info[0],info[1]))

}


dp[0] = 0


for i in 1...c+100 {
    
    for (cost,customer) in arr {
        
        if customer <= i {
            dp[i] = min(dp[i],dp[i-customer] + cost)
        }
    }
    
}

print(dp[c...c+100].min()!) //적어도 c명
